source/grid/grid_generator_from_name.cc

```diff
@@ -0,0 +1,331 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2019 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#include <deal.II/base/mutable_bind.h>
+
+#include <deal.II/grid/grid_generator.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace GridGenerator
+{
+  namespace
+  {
+    template <int dim, int spacedim, class... Arguments>
+    void
+    parse_and_create(void (*generator)(Triangulation<dim, spacedim> &,
+                                       Arguments...),
+                     const std::string &           arguments,
+                     Triangulation<dim, spacedim> &tria)
+    {
+      std::function<void(Arguments...)> wrapper =
+        [&tria, &generator](Arguments... args) { generator(tria, args...); };
+      auto bound_function = Utilities::mutable_bind(wrapper);
+      bound_function.parse_arguments(arguments);
+      bound_function();
+    }
+
+
+    /**
+     * Create grids that only exist for codimension zero combinations.
+     */
+    template <int dim, int spacedim>
+    typename std::enable_if<dim != spacedim, bool>::type
+    generate_codimension_zero_grid(const std::string &,
+                                   const std::string &,
+                                   Triangulation<dim, spacedim> &)
+    {
+      return false;
+    }
+
+    /**
+     * Create grids that only exist for codimension zero combinations.
+     */
+    template <int dim>
+    bool
+    generate_codimension_zero_grid(const std::string & name,
+                                   const std::string & arguments,
+                                   Triangulation<dim> &tria)
+    {
+      if (name == "simplex")
+        parse_and_create<dim, dim, const std::vector<Point<dim>> &>(simplex,
+                                                                    arguments,
+                                                                    tria);
+      else if (name == "subdivided_hyper_rectangle")
+        {
+          try
+            {
+              parse_and_create<dim,
+                               dim,
+                               const std::vector<unsigned int> &,
+                               const Point<dim> &,
+                               const Point<dim> &,
+                               bool>(subdivided_hyper_rectangle,
+                                     arguments,
+                                     tria);
+            }
+          catch (...)
```

Comment:
Can you please document why this is needed? Would it make sense to only catch the specific exception happening here?